Default cargo publish to the alt registry if it's the only allowed one

canova opened the pull request:
Hi, this PR fixes #8036. Previously if we had only one allowed registry in Cargo.toml file, `cargo publish` was failing. But with this PR, we are defaulting to the only allowed registry with printing a note if there is only one registry in that array. I believe this will make things easier for people who use an alternative registry all the time.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Eh2406 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

ehuss commented:
Thanks for the PR!  Would you mind updating the documentation to take note of the behavior?  I would add a small note to the bottom of the section https://github.com/rust-lang/cargo/blob/master/src/doc/src/reference/manifest.md#the-publish-field. And also update the `--registry` flag documentation for the `cargo publish` man page.  The man pages have recently been rearranged, so you'll need to rebase your PR to the lastest master.  The source is in https://github.com/rust-lang/cargo/blob/master/src/doc/man/cargo-publish.md.  I would remove `{{> options-registry }}` which loads a generic template for the `--registry` flag and add a custom definition, that might look something like this:

```
{{#option "`--registry` _registry_"}}
Name of the registry to publish to. Registry names are defined in [Cargo
config files](../reference/config.html). If not specified, and there is a
[`package.publish`](../reference/manifest.html#the-publish-field) field in
`Cargo.toml` with a single registry, then it will publish to that registry.
Otherwise it will use the default registry, which is defined by the
[`registry.default`](../reference/config.html#registry-default) config key
which defaults to `crates-io`.
{{/option}}
```

You can then run the [`build-man.sh`](https://github.com/rust-lang/cargo/blob/master/src/doc/build-man.sh) script to rebuild the man pages.

canova commented:
Thanks @ehuss, updated the code! Rebased on top of current master and added two more commits for the doc/man page update you mentioned.  Could you take a look?

ehuss commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit a8ce734a95a1450882550a66d2044560e989bbcc has been approved by `ehuss`

<!-- @bors r=ehuss a8ce734a95a1450882550a66d2044560e989bbcc -->
<!-- homu: {"type":"Approved","sha":"a8ce734a95a1450882550a66d2044560e989bbcc","approver":"ehuss"} -->

bors commented:
:hourglass: Testing commit a8ce734a95a1450882550a66d2044560e989bbcc with merge 449743b1ac084b67e11fafb278169291f15fcfc3...
<!-- homu: {"type":"BuildStarted","head_sha":"a8ce734a95a1450882550a66d2044560e989bbcc","merge_sha":"449743b1ac084b67e11fafb278169291f15fcfc3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/955105341)
Approved by: ehuss
Pushing 449743b1ac084b67e11fafb278169291f15fcfc3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/955105341"},"merge_sha":"449743b1ac084b67e11fafb278169291f15fcfc3"} -->
